src/content/post/20161129_git_reflog_amend.md

```diff
@@ -0,0 +1,114 @@
+
+But you can also do this:
+
+<pre>
+# Add new line and prepare them for the commit.
+git add -p
+# Amend the previous commit and add to it the staged modifications.
+git commit --amend
+</pre>
+
+A this point, you are asked to edit the previous commit message and the staged edits will be included in it. Useful, isnt'it?
+
+## The side effect
+
+This little trick can be very handy in many situations, but it can also be a very sharp double-edge blade because, as you probably have guessed, you can use the amend on *any* commit including, for example, a *merge commit*.
+
+So, here's a possible sequence of commands that could generate a bad situation.
+
+<pre>
+# You add some modifications.
+git add -p
+# Then create a new commit
+git commit -m 'Added new public method'
+# Then merge the last develop in your working branch
+git fetch --all
+git merge origin/develop
+</pre>
+
+So far the history of your branch could be something like:
+
+<pre>
+~$ git log --oneline
+
+b1748d6 Merge remote-tracking branch 'origin/develop' into feature/my_branch
+a69fa15 Added new public method
+</pre>
+
+Now you accidendally amend the next modifications instead of creating a new commit.
+
+<pre>
+# Write a block comment for your new method and stage the new modifcations.
+git add -p
+# But instead of creating a new commit, you use the --amend option.
+git commit --amend
+</pre>
+
+At this point, the history will be almost the same, but the hash of the *merge commit* will be different and that commmit will includeyour new block comment too.
+
+<pre>
+~$ git log --oneline
+
+f5a5377 Merge remote-tracking branch 'origin/develop' into feature/my_branch
+a69fa15 Added new public method
+</pre>
+
+That's bad.
+
+## The git reflog solution
+
+Because you want to restore the situation before the amend  you can decide to use the `git reflog` in combination with a `git reset`.
+
+The `git reflog` command will prompt the history of the last commands you've entered.
+The output will look something like this.
+
+<pre>
+~$ git reflog
+
+f5a5377 HEAD@{0}: commit (amend): Merge remote-tracking branch 'origin/develop' into feature/my_branch
+b1748d6 HEAD@{1}: merge origin/develop: Merge made by the 'recursive' strategy.
+a69fa15 HEAD@{2}: Added new public method
+</pre>
+
+As you can see you have a separate line for each commit action made, and it is pretty easy to solve the problem. Having the **HEAD@{1}** hash now you can go back to there  *de facto* removing the last command from the history.
+
+<pre>
+# We go back in the history up to the merge commit.
+git reset --soft b1748d6
+# At this point we'll have all the edits that we wrongly added to the merge commit already staged and ready to be added to a new commit.
```

Comment:
What about amending a commit BEFORE the merge? Is this possible? Would you mind putting information as an added bonus to the article? Or am I totally missing the point in having a clean history?

Would be nice.